From c64a510f50290baa94b518f8e75adc6f46d7053d Mon Sep 17 00:00:00 2001
From: Alex Perrin <44126061+AlexPerrin@users.noreply.github.com>
Date: Sat, 17 Oct 2026 14:47:16 -0400
Subject: [PATCH] Grand Exchange: trade history

The clerks' History op and the GE interface's History button open ge_history (383),
filled with ge_history_init / ge_history_addline / ge_history_finish (1644-1646):
the last 40 finished offers (and aborted ones with something traded), newest first,
with the sale tax. A line's Buy-offer starts a buy offer for its item in the first free
slot; the Exchange button goes back. History lives in 160 perm varps.
---
 .../pack/src/main/resources/gamevals.toml     | 160 ++++
 .../pack/configs/grand_exchange.toml          | 800 ++++++++++++++++++
 .../grandexchange/GrandExchangeScriptTest.kt  |  42 +
 .../interfaces/grandexchange/GrandExchange.kt |  47 +-
 .../grandexchange/GrandExchangeScript.kt      |  61 ++
 5 files changed, 1108 insertions(+), 2 deletions(-)

diff --git a/content/interfaces/grand-exchange/pack/src/main/resources/gamevals.toml b/content/interfaces/grand-exchange/pack/src/main/resources/gamevals.toml
index 63e1de86e..87f8b3e3a 100644
--- a/content/interfaces/grand-exchange/pack/src/main/resources/gamevals.toml
+++ b/content/interfaces/grand-exchange/pack/src/main/resources/gamevals.toml
@@ -96,6 +96,166 @@ ge_limit14_bought = 64029
 ge_limit15_item = 64030
 ge_limit15_start = 64031
 ge_limit15_bought = 64032
+ge_history0_obj = 63000
+ge_history0_count = 63001
+ge_history0_gold = 63002
+ge_history0_tax = 63003
+ge_history1_obj = 63004
+ge_history1_count = 63005
+ge_history1_gold = 63006
+ge_history1_tax = 63007
+ge_history2_obj = 63008
+ge_history2_count = 63009
+ge_history2_gold = 63010
+ge_history2_tax = 63011
+ge_history3_obj = 63012
+ge_history3_count = 63013
+ge_history3_gold = 63014
+ge_history3_tax = 63015
+ge_history4_obj = 63016
+ge_history4_count = 63017
+ge_history4_gold = 63018
+ge_history4_tax = 63019
+ge_history5_obj = 63020
+ge_history5_count = 63021
+ge_history5_gold = 63022
+ge_history5_tax = 63023
+ge_history6_obj = 63024
+ge_history6_count = 63025
+ge_history6_gold = 63026
+ge_history6_tax = 63027
+ge_history7_obj = 63028
+ge_history7_count = 63029
+ge_history7_gold = 63030
+ge_history7_tax = 63031
+ge_history8_obj = 63032
+ge_history8_count = 63033
+ge_history8_gold = 63034
+ge_history8_tax = 63035
+ge_history9_obj = 63036
+ge_history9_count = 63037
+ge_history9_gold = 63038
+ge_history9_tax = 63039
+ge_history10_obj = 63040
+ge_history10_count = 63041
+ge_history10_gold = 63042
+ge_history10_tax = 63043
+ge_history11_obj = 63044
+ge_history11_count = 63045
+ge_history11_gold = 63046
+ge_history11_tax = 63047
+ge_history12_obj = 63048
+ge_history12_count = 63049
+ge_history12_gold = 63050
+ge_history12_tax = 63051
+ge_history13_obj = 63052
+ge_history13_count = 63053
+ge_history13_gold = 63054
+ge_history13_tax = 63055
+ge_history14_obj = 63056
+ge_history14_count = 63057
+ge_history14_gold = 63058
+ge_history14_tax = 63059
+ge_history15_obj = 63060
+ge_history15_count = 63061
+ge_history15_gold = 63062
+ge_history15_tax = 63063
+ge_history16_obj = 63064
+ge_history16_count = 63065
+ge_history16_gold = 63066
+ge_history16_tax = 63067
+ge_history17_obj = 63068
+ge_history17_count = 63069
+ge_history17_gold = 63070
+ge_history17_tax = 63071
+ge_history18_obj = 63072
+ge_history18_count = 63073
+ge_history18_gold = 63074
+ge_history18_tax = 63075
+ge_history19_obj = 63076
+ge_history19_count = 63077
+ge_history19_gold = 63078
+ge_history19_tax = 63079
+ge_history20_obj = 63080
+ge_history20_count = 63081
+ge_history20_gold = 63082
+ge_history20_tax = 63083
+ge_history21_obj = 63084
+ge_history21_count = 63085
+ge_history21_gold = 63086
+ge_history21_tax = 63087
+ge_history22_obj = 63088
+ge_history22_count = 63089
+ge_history22_gold = 63090
+ge_history22_tax = 63091
+ge_history23_obj = 63092
+ge_history23_count = 63093
+ge_history23_gold = 63094
+ge_history23_tax = 63095
+ge_history24_obj = 63096
+ge_history24_count = 63097
+ge_history24_gold = 63098
+ge_history24_tax = 63099
+ge_history25_obj = 63100
+ge_history25_count = 63101
+ge_history25_gold = 63102
+ge_history25_tax = 63103
+ge_history26_obj = 63104
+ge_history26_count = 63105
+ge_history26_gold = 63106
+ge_history26_tax = 63107
+ge_history27_obj = 63108
+ge_history27_count = 63109
+ge_history27_gold = 63110
+ge_history27_tax = 63111
+ge_history28_obj = 63112
+ge_history28_count = 63113
+ge_history28_gold = 63114
+ge_history28_tax = 63115
+ge_history29_obj = 63116
+ge_history29_count = 63117
+ge_history29_gold = 63118
+ge_history29_tax = 63119
+ge_history30_obj = 63120
+ge_history30_count = 63121
+ge_history30_gold = 63122
+ge_history30_tax = 63123
+ge_history31_obj = 63124
+ge_history31_count = 63125
+ge_history31_gold = 63126
+ge_history31_tax = 63127
+ge_history32_obj = 63128
+ge_history32_count = 63129
+ge_history32_gold = 63130
+ge_history32_tax = 63131
+ge_history33_obj = 63132
+ge_history33_count = 63133
+ge_history33_gold = 63134
+ge_history33_tax = 63135
+ge_history34_obj = 63136
+ge_history34_count = 63137
+ge_history34_gold = 63138
+ge_history34_tax = 63139
+ge_history35_obj = 63140
+ge_history35_count = 63141
+ge_history35_gold = 63142
+ge_history35_tax = 63143
+ge_history36_obj = 63144
+ge_history36_count = 63145
+ge_history36_gold = 63146
+ge_history36_tax = 63147
+ge_history37_obj = 63148
+ge_history37_count = 63149
+ge_history37_gold = 63150
+ge_history37_tax = 63151
+ge_history38_obj = 63152
+ge_history38_count = 63153
+ge_history38_gold = 63154
+ge_history38_tax = 63155
+ge_history39_obj = 63156
+ge_history39_count = 63157
+ge_history39_gold = 63158
+ge_history39_tax = 63159
 
 [gamevals.enum]
 ge_buy_limits = 65380
diff --git a/content/interfaces/grand-exchange/pack/src/main/resources/pack/configs/grand_exchange.toml b/content/interfaces/grand-exchange/pack/src/main/resources/pack/configs/grand_exchange.toml
index be5b35798..30feb177a 100644
--- a/content/interfaces/grand-exchange/pack/src/main/resources/pack/configs/grand_exchange.toml
+++ b/content/interfaces/grand-exchange/pack/src/main/resources/pack/configs/grand_exchange.toml
@@ -532,6 +532,806 @@ id = "varp.ge_limit15_bought"
 scope = "Perm"
 transmit = "Never"
 
+[[varp]]
+id = "varp.ge_history0_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history0_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history0_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history0_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history1_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history1_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history1_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history1_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history2_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history2_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history2_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history2_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history3_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history3_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history3_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history3_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history4_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history4_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history4_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history4_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history5_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history5_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history5_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history5_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history6_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history6_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history6_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history6_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history7_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history7_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history7_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history7_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history8_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history8_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history8_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history8_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history9_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history9_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history9_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history9_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history10_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history10_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history10_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history10_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history11_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history11_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history11_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history11_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history12_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history12_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history12_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history12_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history13_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history13_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history13_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history13_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history14_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history14_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history14_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history14_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history15_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history15_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history15_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history15_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history16_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history16_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history16_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history16_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history17_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history17_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history17_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history17_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history18_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history18_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history18_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history18_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history19_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history19_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history19_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history19_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history20_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history20_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history20_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history20_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history21_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history21_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history21_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history21_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history22_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history22_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history22_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history22_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history23_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history23_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history23_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history23_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history24_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history24_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history24_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history24_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history25_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history25_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history25_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history25_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history26_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history26_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history26_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history26_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history27_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history27_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history27_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history27_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history28_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history28_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history28_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history28_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history29_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history29_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history29_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history29_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history30_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history30_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history30_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history30_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history31_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history31_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history31_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history31_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history32_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history32_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history32_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history32_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history33_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history33_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history33_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history33_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history34_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history34_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history34_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history34_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history35_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history35_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history35_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history35_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history36_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history36_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history36_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history36_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history37_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history37_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history37_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history37_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history38_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history38_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history38_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history38_tax"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history39_obj"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history39_count"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history39_gold"
+scope = "Perm"
+transmit = "Never"
+
+[[varp]]
+id = "varp.ge_history39_tax"
+scope = "Perm"
+transmit = "Never"
+
 [[inventory]]
 isServerOnly = true
 id = "inv.tradingpost_sell_0"
diff --git a/content/interfaces/grand-exchange/src/integration/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchangeScriptTest.kt b/content/interfaces/grand-exchange/src/integration/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchangeScriptTest.kt
index 2ff6a066a..2d61696b6 100644
--- a/content/interfaces/grand-exchange/src/integration/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchangeScriptTest.kt
+++ b/content/interfaces/grand-exchange/src/integration/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchangeScriptTest.kt
@@ -45,6 +45,11 @@ class GrandExchangeScriptTest {
             GrandExchange.clear(player, slot)
             player.clearInv(GrandExchange.collectionInv(player, slot))
         }
+        for (i in 0 until GrandExchange.HISTORY_SIZE) {
+            for (field in listOf("obj", "count", "gold", "tax")) {
+                GrandExchange.set(player, "varp.ge_history${i}_$field", 0)
+            }
+        }
         for (i in 0 until GrandExchange.LIMIT_TRACKERS) {
             for (field in listOf("item", "start", "bought")) {
                 GrandExchange.set(player, "varp.ge_limit${i}_$field", 0)
@@ -292,6 +297,43 @@ class GrandExchangeScriptTest {
             )
         }
 
+    @Test
+    fun GameTestState.`finished and partly filled aborted offers go into the history, newest first`() =
+        runInjectedGameTest(GrandExchangeScript::class, null, GrandExchangeScript::class) { script ->
+            resetGe(player)
+            player.inv[0] = InvObj(id("obj.coins"), 100_000)
+            repeat(3) { player.inv[1 + it] = InvObj(lobster.id, 1) }
+            with(script) {
+                player.withProtectedAccess {
+                    setUp(slot = 0, sell = false, item = lobster.id, count = 10, price = guide)
+                    assertTrue(confirm())
+                    setUp(slot = 1, sell = true, item = lobster.id, count = 3, price = 1)
+                    assertTrue(confirm())
+                    setUp(slot = 2, sell = false, item = lobster.id, count = 5, price = 1)
+                    assertTrue(confirm())
+                }
+            }
+            GrandExchange.fillOpenOffers(player, NOW) { it.cost }
+            assertTrue(GrandExchange.abort(player, 0))
+            assertTrue(GrandExchange.abort(player, 2))
+            fillAll(player)
+            val history = GrandExchange.history(player)
+            assertEquals(2, history.size)
+            assertEquals(GrandExchange.Trade(lobster.id, true, 3, 3 * guide, 3 * GrandExchange.taxPerItem(guide)), history[0])
+            assertEquals(GrandExchange.Trade(lobster.id, false, 3, 3 * guide, 0), history[1])
+
+            repeat(GrandExchange.HISTORY_SIZE) {
+                GrandExchange.recordHistory(
+                    player,
+                    GrandExchange.Offer(0, id("obj.coal"), 1, 1, 1, it + 1, GrandExchange.STATE_FINISHED),
+                )
+            }
+            val full = GrandExchange.history(player)
+            assertEquals(GrandExchange.HISTORY_SIZE, full.size)
+            assertEquals(GrandExchange.HISTORY_SIZE, full.first().gold)
+            assertEquals(1, full.last().gold)
+        }
+
     private companion object {
         const val NOW = 29_000_000
     }
diff --git a/content/interfaces/grand-exchange/src/main/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchange.kt b/content/interfaces/grand-exchange/src/main/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchange.kt
index 5d5fe4f08..9f1a21864 100644
--- a/content/interfaces/grand-exchange/src/main/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchange.kt
+++ b/content/interfaces/grand-exchange/src/main/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchange.kt
@@ -56,6 +56,10 @@ object GrandExchange {
     /** How often a limit-blocked offer looks again (100 ticks = 1 minute). */
     const val LIMIT_RECHECK = 100
 
+    const val HISTORY_SIZE = 40
+    /** Bit set in `ge_historyN_obj` for a sale (the low bits hold obj + 1). */
+    private const val HISTORY_SOLD_BIT = 1 shl 30
+
     const val LIMIT_WINDOW_MINUTES = 240
     const val LIMIT_TRACKERS = 16
 
@@ -226,6 +230,41 @@ object GrandExchange {
         set(player, limitVarp(index, "bought"), count)
     }
 
+    data class Trade(val item: Int, val sold: Boolean, val count: Int, val gold: Int, val tax: Int)
+
+    private fun historyVarp(index: Int, field: String) = "varp.ge_history${index}_$field"
+
+    /** The trade history, newest first. */
+    fun history(player: Player): List<Trade> =
+        (0 until HISTORY_SIZE).mapNotNull { i ->
+            val obj = player.vars[historyVarp(i, "obj")]
+            if (obj == 0) return@mapNotNull null
+            Trade(
+                item = (obj and HISTORY_SOLD_BIT.inv()) - 1,
+                sold = obj and HISTORY_SOLD_BIT != 0,
+                count = player.vars[historyVarp(i, "count")],
+                gold = player.vars[historyVarp(i, "gold")],
+                tax = player.vars[historyVarp(i, "tax")],
+            )
+        }
+
+    /** Adds a finished offer with at least one item traded to the front of the history. */
+    fun recordHistory(player: Player, offer: Offer) {
+        if (offer.done <= 0) return
+        val tax = if (offer.sell) offer.done * taxPerItem(offer.gold / offer.done) else 0
+        val entries =
+            listOf(Trade(offer.item, offer.sell, offer.done, offer.gold, tax)) +
+                history(player).take(HISTORY_SIZE - 1)
+        for (i in 0 until HISTORY_SIZE) {
+            val trade = entries.getOrNull(i)
+            val obj = trade?.let { (it.item + 1) or (if (it.sold) HISTORY_SOLD_BIT else 0) } ?: 0
+            set(player, historyVarp(i, "obj"), obj)
+            set(player, historyVarp(i, "count"), trade?.count ?: 0)
+            set(player, historyVarp(i, "gold"), trade?.gold ?: 0)
+            set(player, historyVarp(i, "tax"), trade?.tax ?: 0)
+        }
+    }
+
     fun stepSize(count: Int): Int = ((count + FILL_STEPS - 1) / FILL_STEPS).coerceAtLeast(1)
 
     sealed interface StepResult {
@@ -277,8 +316,10 @@ object GrandExchange {
         val done = offer.done + amount
         val finished = done >= offer.count
         val status = if (finished) finishedStatus(offer.sell) else offer.status
-        store(player, offer.copy(done = done, gold = offer.gold + gold, status = status))
+        val updated = offer.copy(done = done, gold = offer.gold + gold, status = status)
+        store(player, updated)
         sendSlot(player, offer.slot)
+        if (finished) recordHistory(player, updated)
         val verb = if (offer.sell) "selling" else "buying"
         if (finished) {
             player.mes("Grand Exchange: Finished $verb ${offer.count.commas()} x ${type.name}.")
@@ -299,8 +340,10 @@ object GrandExchange {
         } else {
             addTo(inv, 1, coinsId, (offer.count - offer.done) * offer.price)
         }
-        store(player, offer.copy(status = finishedStatus(offer.sell)))
+        val aborted = offer.copy(status = finishedStatus(offer.sell))
+        store(player, aborted)
         sendSlot(player, slot)
+        recordHistory(player, aborted)
         return true
     }
 
diff --git a/content/interfaces/grand-exchange/src/main/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchangeScript.kt b/content/interfaces/grand-exchange/src/main/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchangeScript.kt
index 193d0d9f5..d0bd96378 100644
--- a/content/interfaces/grand-exchange/src/main/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchangeScript.kt
+++ b/content/interfaces/grand-exchange/src/main/kotlin/org/rsmod/content/interfaces/grandexchange/GrandExchangeScript.kt
@@ -16,7 +16,9 @@ import org.rsmod.api.script.onIfClose
 import org.rsmod.api.script.onIfModalButton
 import org.rsmod.api.script.onIfOpen
 import org.rsmod.api.script.onApNpc3
+import org.rsmod.api.script.onApNpc4
 import org.rsmod.api.script.onOpNpc3
+import org.rsmod.api.script.onOpNpc4
 import org.rsmod.api.script.onPlayerLogin
 import org.rsmod.api.script.onPlayerSoftTimer
 import org.rsmod.content.interfaces.grandexchange.GrandExchange.commas
@@ -35,6 +37,8 @@ class GrandExchangeScript @Inject constructor(private val marketPrices: MarketPr
         for (clerk in CLERKS) {
             onOpNpc3(clerk) { openExchange() }
             onApNpc3(clerk) { if (isWithinApRange(it.npc, distance = 2)) openExchange() }
+            onOpNpc4(clerk) { openHistory() }
+            onApNpc4(clerk) { if (isWithinApRange(it.npc, distance = 2)) openHistory() }
         }
 
         onPlayerLogin {
@@ -63,6 +67,9 @@ class GrandExchangeScript @Inject constructor(private val marketPrices: MarketPr
             }
         }
         onIfModalButton("component.ge_offers:back") { backToIndex() }
+        onIfModalButton("component.ge_offers:history") { openHistory() }
+        onIfModalButton("component.ge_history:exchange") { openExchange() }
+        onIfModalButton("component.ge_history:list") { historyOp(it.comsub, it.op) }
         onIfModalButton("component.ge_offers:setup") { setupOp(it.comsub, it.op) }
         onIfModalButton("component.ge_offers:setup_confirm") { confirm() }
         onIfModalButton("component.ge_offers:details_status") {
@@ -129,6 +136,48 @@ class GrandExchangeScript @Inject constructor(private val marketPrices: MarketPr
         ifSetEvents("component.ge_offers_side:items", inv.indices, IfEvent.Op1, IfEvent.Op10)
     }
 
+    /** The trade history (`ge_history` 383), drawn by scripts 1644-1646, newest first. */
+    private fun ProtectedAccess.openHistory() {
+        ifOpenMain("interface.ge_history")
+        player.runClientScript(HISTORY_INIT_SCRIPT)
+        GrandExchange.history(player).forEachIndexed { i, trade ->
+            player.runClientScript(
+                HISTORY_ADDLINE_SCRIPT,
+                i,
+                trade.item,
+                if (trade.sold) 1 else 0,
+                trade.count,
+                trade.gold.toLong(),
+                trade.tax.toLong(),
+            )
+        }
+        player.runClientScript(HISTORY_FINISH_SCRIPT)
+        ifSetEvents(
+            "component.ge_history:list",
+            0 until GrandExchange.HISTORY_SIZE * HISTORY_LINE_COMPONENTS,
+            IfEvent.Op1,
+            IfEvent.Op10,
+        )
+    }
+
+    /** "Buy-offer" on a history line opens a buy offer for its item; "Examine" examines it. */
+    private fun ProtectedAccess.historyOp(comsub: Int, op: IfButtonOp) {
+        val trade = GrandExchange.history(player).getOrNull(comsub / HISTORY_LINE_COMPONENTS) ?: return
+        val type = GrandExchange.type(trade.item) ?: return
+        if (op == IfButtonOp.Op10) {
+            mes(type.examine)
+            return
+        }
+        val slot = firstFreeSlot()
+        openExchange()
+        if (slot == null) {
+            mes("You have no free offer slots.")
+            return
+        }
+        resetSetup(player, slot + 1)
+        selectItem(type, quantity = 1)
+    }
+
     private fun setCollectionBoxEvents(player: Player) {
         for (slot in 0 until GrandExchange.SLOTS) {
             player.ifSetEvents(
@@ -409,6 +458,18 @@ class GrandExchangeScript @Inject constructor(private val marketPrices: MarketPr
         val QUANTITY_DELTAS = mapOf(1 to -1, 2 to 1, 3 to 1, 4 to 10, 5 to 100, 6 to Int.MAX_VALUE)
         val PRICE_DELTAS = mapOf(8 to -1, 9 to 1, 10 to -5, 13 to 5)
 
+        /** Components `ge_history_addline` creates per line (its sub ids are line * 6 + n). */
+        const val HISTORY_LINE_COMPONENTS = 6
+        val HISTORY_INIT_SCRIPT: Int by lazy {
+            "clientscript.[clientscript,ge_history_init]".asRSCM(RSCMType.CLIENTSCRIPT)
+        }
+        val HISTORY_ADDLINE_SCRIPT: Int by lazy {
+            "clientscript.[clientscript,ge_history_addline]".asRSCM(RSCMType.CLIENTSCRIPT)
+        }
+        val HISTORY_FINISH_SCRIPT: Int by lazy {
+            "clientscript.[clientscript,ge_history_finish]".asRSCM(RSCMType.CLIENTSCRIPT)
+        }
+
         val SETDESC_SCRIPT: Int by lazy {
             "clientscript.[clientscript,ge_offers_setdesc]".asRSCM(RSCMType.CLIENTSCRIPT)
         }
-- 
2.43.0

